stringdtype/stringdtype/src/casts.c

```diff
@@ -64,12 +65,293 @@ PyArrayMethod_Spec StringToStringCastSpec = {
+                  NpyAuxData *NPY_UNUSED(auxdata))
+{
+    PyArray_Descr **descrs = context->descriptors;
+    long in_size = (descrs[0]->elsize) / 4;
+
+    npy_intp N = dimensions[0];
+    Py_UCS4 *in = (Py_UCS4 *)data[0];
+    char **out = (char **)data[1];
+
+    // 4 bytes per UCS4 character
+    npy_intp in_stride = strides[0] / 4;
+    // strides are in bytes but pointer offsets are in pointer widths, so
+    // divide by the element size (one pointer width) to get the pointer offset
+    npy_intp out_stride = strides[1] / context->descriptors[1]->elsize;
+
+    while (N--) {
+        // pessimistically allocate 4 bytes per allowed character
+        // plus one byte for the null terminator
+        char *out_buf = malloc((in_size * 4 + 1) * sizeof(char));
+        size_t out_num_bytes = 0;
+        for (int i = 0; i < in_size; i++) {
+            // get code point
+            Py_UCS4 code = in[i];
+
+            if (code == 0) {
+                break;
+            }
+
+            // convert codepoint to UTF8 bytes
+            char utf8_c[4] = {0};
+            size_t num_bytes = ucs4_code_to_utf8_char(code, utf8_c);
+            out_num_bytes += num_bytes;
+
+            if (num_bytes == -1) {
+                // acquire GIL, set error, return
+                PyGILState_STATE gstate;
+                gstate = PyGILState_Ensure();
+                PyErr_SetString(PyExc_TypeError,
+                                "Invalid unicode code point found");
+                PyGILState_Release(gstate);
+                return -1;
+            }
+
+            // copy utf8_c into out_buf
+            strncpy(out_buf, utf8_c, num_bytes);
+
+            // increment out_buf by the size of the character
+            out_buf += num_bytes;
+        }
+
+        // reset out_buf to the beginning of the string
+        out_buf -= out_num_bytes;
+
+        // pad string with null character
+        out_buf[out_num_bytes] = '\0';
+
+        // resize out_buf now that we know the real size
+        out_buf = realloc(out_buf, out_num_bytes + 1);
+
+        // set out to the address of the beginning of the string
+        out[0] = out_buf;
+
+        in += in_stride;
+        out += out_stride;
+    }
+
+    return 0;
+}
+
+static PyType_Slot u2s_slots[] = {
+        {NPY_METH_resolve_descriptors, &unicode_to_string_resolve_descriptors},
+        {NPY_METH_strided_loop, &unicode_to_string},
+        {0, NULL}};
+
+static char *u2s_name = "cast_Unicode_to_StringDType";
+
+static NPY_CASTING
+string_to_unicode_resolve_descriptors(PyObject *NPY_UNUSED(self),
+                                      PyArray_DTypeMeta *NPY_UNUSED(dtypes[2]),
+                                      PyArray_Descr *given_descrs[2],
+                                      PyArray_Descr *loop_descrs[2],
+                                      npy_intp *NPY_UNUSED(view_offset))
+{
+    Py_INCREF(given_descrs[0]);
+    loop_descrs[0] = given_descrs[0];
```

Comment:
Move this to the end/after the error, so you don't have to worry about cleanup when the error path is hit.